### PR TITLE
refine impl of rename file/directory

### PR DESCRIPTION
**Pull request by Jinming-Hu**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-files-datalake/src/datalake_directory_client.cpp [hunk 1]
```diff
@@ -110,7 +110,7 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
     protocolLayerOptions.SourceIfModifiedSince = options.SourceAccessConditions.IfModifiedSince;
     protocolLayerOptions.SourceIfUnmodifiedSince = options.SourceAccessConditions.IfUnmodifiedSince;
     protocolLayerOptions.RenameSource
-        = "/" + m_pathUrl.GetPath() + _internal::UrlEncodePath(fileName);
+        = "/" + m_pathUrl.GetPath() + "/" + _internal::UrlEncodePath(fileName);
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-datalake/src/datalake_directory_client.cpp:113]: Do we have tests that would catch this? If not, add tests.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/src/datalake_directory_client.cpp:113]: there's a bug here. Working on it. will add tests

**Review by RickWinter — APPROVED** [review]

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_utilities.hpp [hunk 2]
```diff
@@ -12,8 +12,16 @@
...
+  Azure::Core::Url GetBlobUrlFromUrl(const Azure::Core::Url& url);
+  Azure::Core::Url GetDfsUrlFromUrl(const Azure::Core::Url& url);
+  inline std::string GetBlobUrlFromUrl(const std::string& url)
+  {
+    return GetBlobUrlFromUrl(Azure::Core::Url(url)).GetAbsoluteUrl();
+  }
+  inline std::string GetDfsUrlFromUrl(const std::string& url)
+  {
+    return GetDfsUrlFromUrl(Azure::Core::Url(url)).GetAbsoluteUrl();
+  }
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_utilities.hpp:24]: I don't understand what these are doing, especially when the names of `Url` accepting ones are identical to the one accepting `std::string`. Add some comments.

File: sdk/storage/azure-storage-files-datalake/src/datalake_utilities.cpp [hunk 3]
```diff
@@ -13,25 +13,31 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake { nam
...
   const static std::string BlobEndPointIdentifier = ".blob.";
 
-  std::string GetBlobUrlFromUrl(const std::string& url)
+  Azure::Core::Url GetBlobUrlFromUrl(const Azure::Core::Url& url)
   {
-    std::string result = url;
-    auto pos = result.find(DfsEndPointIdentifier);
-    if (pos != std::string::npos)
+    std::string host = url.GetHost();
+    auto pos = host.rfind(DfsEndPointIdentifier);
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-datalake/src/datalake_utilities.cpp:19]: Why `rfind`?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_utilities.hpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/datalake_utilities.hpp:24]: It's overloaded functions intended to transform between dfs endpoints and blob endpoints.

Like `someaccount.dfs.core.windows.net` <-> `someaccount.blob.core.windows.net`.

They're in detail namespace.

**Pull request merged**